### ISSUE TITLE
Improve handling of tasks failing/succeeding on unexpected workers

### ISSUE THREAD

**Issue by hendrikmakait** [dask/distributed#6956]
This issue is a follow-up on an offline discussion centering around https://github.com/dask/distributed/pull/6884#issuecomment-1224775494 and #6939 with @crusaderky and @gjoseph92 where we encountered some issues with the way we currently handle tasks failing/succeeding on unexpected workers.

In `transition_processing_memory`, when encountering an unexpected worker that has successfully completed a task, we send a `cancel-compute` message to the worker that was supposed to process this task assuming that the unexpected worker would hold onto our result for us.

https://github.com/dask/distributed/blob/c15a10e87ca5d03e62f0ad4f38adb63163522979/distributed/scheduler.py#L1987-L2001

However, this series of events _should_ only be triggered in a chain of events where `free-keys` message for that task should already be on its way to the unexpected worker, causing it to remove the result. This leaves us with no worker holding on to the task/result.

While we do suspect this chain of events to unfold, we need a test verifying it. If this is indeed the case, we should not send the `cancel-compute` message to `ts.processing_on`, but rather make sure that we clean up anything related to the work happening on the unexpected worker.

In `stimulus_task_erred`, we don't check whether the task `erred` on an unexpected worker at all and run through the retry-logic. In https://github.com/dask/distributed/pull/6884#issuecomment-1217824082, @fjetter suggested instead ignoring that the task `erred` on an unexpected worker and continuing on as if nothing happened. This approach is reasonable as the unexpected worker suggests something faulty going on. We would like to check whether we need to do additional work to ensure that the unexpected worker is reset and forgets about the task and log a warning.

**Comment by gjoseph92** [comment]
Especially since we know https://github.com/dask/distributed/issues/6390 is a problem, I'd like to see this happen (ignore `task-finished` or `task-erred` from unexpected workers, and instead tell them to `free-keys`).

It's not quite the same case as what's mentioned here, but we should think about it as well. This behavior may not be good either: https://github.com/dask/distributed/blob/c15a10e87ca5d03e62f0ad4f38adb63163522979/distributed/scheduler.py#L1982-L1985

As things stand right now, it's entirely possible to get messages from removed workers, so rather than releasing the entire task, we should probably just ignore the spurious message.

Basically, whether the worker exists or not, let's just ignore its message if it's not the one we expected to be talking to?
